src/checkers/cwe_476.ml

```diff
@@ -1,47 +1,280 @@
+  | Bil.Unknown(_) -> NoAccess
+  | Bil.Ite(if_, then_, else_) ->
+    union_access_triple (contains_unchecked if_ state) (contains_unchecked then_ state) (contains_unchecked else_ state)
+  | Bil.Extract(_,_, exp) -> contains_unchecked exp state
+  | Bil.Concat(exp1, exp2) -> union_access (contains_unchecked exp1 state) (contains_unchecked exp2 state)
+
+(* If an formerly unchecked return value was checked then remove all registers pointing
+   to the source of this return value from state. *)
+let checks_value exp state : State.t =
+  match exp with
+  | Bil.Ite(if_, then_, else_) -> begin
+      match contains_unchecked if_ state with
+      | Access(var) ->
+        (* We filter out all registers with the same generating tid, since we have checked
+           the return value of this source *)
+        State.remove_tid state var
+      | MemAccess(_) (* This is a memory access before checking the return value, so do nothing here. *)
+      | NoAccess -> state
+    end
+  | _ -> state
+
+type cwe_hits_type = Tid.t list ref
+
+let add_to_hits (cwe_hits:cwe_hits_type) (tid:Tid.t) =
+  match List.find cwe_hits.contents ~f:(fun elem -> elem = tid) with
+  | Some(_) -> ()
+  | None -> (cwe_hits := (tid :: cwe_hits.contents))
+
+(** flags any access (not just memory access) from an unchecked source as a cwe_hit. *)
+
+let flag_any_access exp state ~cwe_hits =
+  match contains_unchecked exp state with
+  | MemAccess(var) | Access(var) ->
+    let tid = State.find_exn state var in
+    add_to_hits cwe_hits tid;
+    State.remove_tid state var
+  | NoAccess -> state
+
+(** flag all unchecked registers as cwe_hits, return empty state *)
+let flag_all_unchecked_registers state ~cwe_hits =
+  let () = List.iter state ~f:(fun (var, tid) ->
+      add_to_hits cwe_hits tid) in
+  []
+
+(** Updates the state depending on the def. If memory is accessed using an unchecked return value,
+    then the access is added to the list of cwe_hits. *)
+let update_state_def def state ~cwe_hits =
+  let (lhs, rhs) = (Def.lhs def, Def.rhs def) in
+  let state = checks_value rhs state in
+  match contains_unchecked rhs state with
+  | MemAccess(var) -> begin (* we found a case of unchecked return value *)
+      let tid = State.find_exn state var in
+      add_to_hits cwe_hits tid;
+      State.remove_tid state var
+    end
+  | Access(var) -> (* taint the lhs as an unchecked return value *)
+    let tid = State.find_exn state var in
+    State.add state lhs tid
+  | NoAccess -> (* no access to an unchecked return value in rhs. Since lhs is overwritten, it cannot be an unchecked return value anymore. *)
+    State.remove_var state lhs
+
+(** Updates the state depending on the jump. On a jump to a function from the function list
+    taint all return registers as unchecked return values. *)
+let update_state_jmp jmp state ~cwe_hits ~function_names ~program ~block ~strict_call_policy =
```

Comment:
this function is rather long, can't you refactor it and break it down into smaller units?

Reply:
The function is a little bit shorter now.